Updates to Regions and Rules

VincentsSin opened the pull request:
Regions: Added remaining regions, sub-regions, and temples to list.

Rules: Various overworld, plus Snowhead and Great Bay temples, locations added w/ logic. Also added `can_clear` definitions for Snowhead, Great Bay, and Stone Tower Temples.

VincentsSin committed:
Updated Regions

Added remaining game regions and temples

VincentsSin committed:
Updated Rules

Added various overworld locations and logic, as well as Snowhead and Great Bay temples.

CelestialKitsune commented on worlds/mm-recomp/Regions.py:
```diff
@@ -10,14 +10,38 @@ class MMRRegionData(NamedTuple):
+    "Path to Mountain Village": MMRRegionData(["Mountain Village"]),
+    "Mountain Village": MMRRegionData(["Twin Islands", "Path to Snowhead"]),
+    "Twin Islands": MMRRegionData(["Goron Village"]),
+    "Goron Village": MMRRegionData([]),
```
I would also add Goron Shrine (the building where the baby goron is) as there will be different access rules based on day 1 or not for later logic modes

VincentsSin replied:
That will be added later on. This was just a first batch that I wrote out and still got more in the works. That said, there's a lot of locations that need catalogued before any rules can be made.

CelestialKitsune commented on worlds/mm-recomp/Regions.py:
```diff
@@ -10,14 +10,38 @@ class MMRRegionData(NamedTuple):
+    "Snowhead Temple": MMRRegionData([]),
+    "Great Bay": MMRRegionData(["Ocean Spider House", "Pirates' Fortress", "Pinnacle Rock", "Zora Cape"]),
+    "Ocean Spider House": MMRRegionData([]),
+    "Pirates' Fortress": MMRRegionData(["Pirates' Fortress (Interior)"]),
```
Should be (Exterior), and add a new region: `Pirates' Fortress Sewers`, for the area accessed by pressing a Goron Switch

VincentsSin replied:
Updated, also with Sewers region routing out to Pirates' Fortress as well.

CelestialKitsune commented on worlds/mm-recomp/Regions.py:
```diff
@@ -10,14 +10,38 @@ class MMRRegionData(NamedTuple):
+    "Ikana Canyon": MMRRegionData(["Secret Shrine", "Beneath the Well", "Ikana Castle", "Stone Tower", "Stone Tower (Inverted)"]),
+    "Secret Shrine": MMRRegionData([]),
+    "Beneath the Well": MMRRegionData([]),
+    "Ikana Castle": MMRRegionData([]),
```
You can also access Beneath the Well from Ikana Castle

CelestialKitsune commented on worlds/mm-recomp/Regions.py:
```diff
@@ -10,14 +10,38 @@ class MMRRegionData(NamedTuple):
+    "Great Bay Temple": MMRRegionData([]),
+    "Road to Ikana": MMRRegionData(["Ikana Graveyard", "Ikana Canyon"]),
+    "Ikana Graveyard": MMRRegionData([]),
+    "Ikana Canyon": MMRRegionData(["Secret Shrine", "Beneath the Well", "Ikana Castle", "Stone Tower", "Stone Tower (Inverted)"]),
```
Stone Tower (Inverted) access is from Stone Tower only

CelestialKitsune commented on worlds/mm-recomp/Rules.py:
```diff
@@ -72,6 +76,36 @@ def get_region_rules(player):
+        "Pirates' Fortress -> Pirates' Fortress (Interior)":
+            lambda state: state.has("Goron Mask", state, player) or state.has("Hookshot", player),
+        "Zora Cape -> Great Bay Temple":
+            lambda state: can_play_song("New Wave Bossa Nova", state, player) and state.has("Hookshot", player) and state.has("Zora Mask, state, player),
```
```suggestion
            lambda state: can_play_song("New Wave Bossa Nova", state, player) and state.has("Hookshot", player) and state.has("Zora Mask", state, player),
```

CelestialKitsune commented on worlds/mm-recomp/Rules.py:
```diff
@@ -72,6 +76,36 @@ def get_region_rules(player):
+        "Termina Field -> Ikana Graveyard":
+            lambda state: can_play_song("Epona's Song", state, player),
+        "Termina Field -> Ikana Canyon":
+            lambda state: can_play_song("Epona's Song", state, player) and state.has("Hookshot", player) and state.has("Garo Mask", player),
```
```suggestion
            lambda state: can_play_song("Epona's Song", state, player) and state.has("Hookshot", player) and (state.has("Garo Mask", player) or state.has("Gibdo Mask", player)),
```

CelestialKitsune commented on worlds/mm-recomp/Rules.py:
```diff
@@ -151,7 +185,7 @@ def get_location_rules(player):
             lambda state: state.has("All-Night Mask", player),
         "Stock Pot Inn Anju and Kafei":
-            lambda state: can_play_song("Epona's Song", state, player) and state.has("Letter to Kafei", player) and state.has("Pendant of Memories", player) and state.has("Hookshot", player) and (state.has("Garo Mask", player) or state.has("Gibdo Mask", player)),
+            lambda state: state.has("Letter to Kafei", player) and state.has("Pendant of Memories", player) and state.has("Hookshot", player) and (state.has("Garo Mask", player) or state.has("Gibdo Mask", player)),
```
You need to be able to complete Sakon's Hideout with Kafei, to trigger this check, so Epona's Song is needed (adding events for the Owl Statues in the future will alleviate this strict need for Epona)

VincentsSin replied:
This is one Cube already had wrote up, so I'll let them resolve

CelestialKitsune replied:
It was modified by this commit though [c309c0](https://github.com/LittlestCube/Archipelago/pull/3/commits/c309c0ffaa4061cf5b120128641fdc850cf0ae33#diff-eb1ca8adc44300e1d89a1de05138b573a8bdeb2938808e0c802f62d4d67b1a66L154-R188)

CelestialKitsune commented on worlds/mm-recomp/Rules.py:
```diff
@@ -296,14 +330,191 @@ def get_location_rules(player):
+        "Mountain Village Darmani":
+            lambda state: state.has("Lens of Truth", state, player) and state.has("Progressive Magic Upgrade", player) and can_play_song("Song of Healing", state, player),
+        "Mountain Village Hungry Goron":
+            lambda state: state.has("Goron Mask", player) and (can_use_fire_arrows(state, player) or (state.has("Deku Stick", state, player) and state.has("Progressive Magic Upgrade", player))),
```
This needs to be changed.
```suggestion
            lambda state: state.has("Goron Mask", player) and state.has("Progressive Magic Upgrade", player) and (can_play_song("Goron's Lullaby", state, player) and state.has("Deku Stick", player) or can_use_fire_arrows(state, player)),
```

LittleCube-hax replied:
Deku Stick isn't an item, so that part of the condition will currently never be true.

CelestialKitsune commented on worlds/mm-recomp/Rules.py:
```diff
@@ -296,14 +330,191 @@ def get_location_rules(player):
+        "Mountain Village Hungry Goron":
+            lambda state: state.has("Goron Mask", player) and (can_use_fire_arrows(state, player) or (state.has("Deku Stick", state, player) and state.has("Progressive Magic Upgrade", player))),
+        "Mountain Village Spring Ramp Grotto":
+            lambda state: state.has("Hookshot", player), # ~ or can_clear_snowhead(state, player),
```
Spring is absolutely required for that check, I would use `can_clear_snowhead(state, player)` if that macro works

VincentsSin replied:
The macro is commented out for now until Goht's Remains location is ID'd into the item pool to be checked.

CelestialKitsune replied:
Well, it needs extra logic then, to ensure you can reach Spring

VincentsSin replied:
Uncommented the condition, since these won't be immediately implemented

CelestialKitsune commented on worlds/mm-recomp/Rules.py:
```diff
@@ -296,14 +330,191 @@ def get_location_rules(player):
+        "Twin Islands Hot Water Grotto Chest":
+            lambda state: has_explosives(state, player) and ((can_use_fire_arrows(state, player) or state.can_reach("Mountain Village Darmani", 'Location', player)), # ~ or can_clear_snowhead(state, player)),
+        "Twin Islands Spring Underwater Cave Chest":
+            lambda state: state.has("Zora Mask", player), # ~ and can_clear_snowhead(state, player)
```
Same comment as for the MV Ramp Grotto

VincentsSin replied:
Same as above

CelestialKitsune commented on worlds/mm-recomp/Rules.py:
```diff
@@ -296,14 +330,191 @@ def get_location_rules(player):
+        "Twin Islands Spring Underwater Cave Chest":
+            lambda state: state.has("Zora Mask", player), # ~ and can_clear_snowhead(state, player)
+        "Twin Islands Spring Underwater Center Chest":
+            lambda state: state.has("Zora Mask", player), # ~ and can_clear_snowhead(state, player)
```
Idem here

VincentsSin replied:
See above

CelestialKitsune commented on worlds/mm-recomp/Rules.py:
```diff
@@ -296,14 +330,191 @@ def get_location_rules(player):
+        "Mountain Title Deed":
+            lambda state: state.has("Swamp Title Deed", player),
+        "Goron Village Freestanding HP":
+            lambda state: state.can_reach("Mountain Title Deed", 'Location', player),
```
Both the Deed and the HP checks need to have Deku Mask

VincentsSin replied:
I swore I had Deku Mask wrote in for this, oops

CelestialKitsune commented on worlds/mm-recomp/Rules.py:
```diff
@@ -296,14 +330,191 @@ def get_location_rules(player):
+        "Great Bay Temple Four Torches Chest":
+            lambda state: can_use_fire_arrows(state, player),
+        "Great Bay Temple Wart Chest":
+            lambda state.has("Progressive Bow", player),
```
```suggestion
            lambda state: state.has("Progressive Bow", player),
```

CelestialKitsune commented on worlds/mm-recomp/Rules.py:
```diff
@@ -296,14 +330,191 @@ def get_location_rules(player):
+            
+            
+        "Graveyard Day 1 Chest":
+            lambda state: state.has("Captain's Hat", player) and can_smack_hard(state, player) and can_use_fire_arrows(state, player),
```
Lighting the torches does not spawn the chest. However it opens the door leading to the Iron Knuckle, that needs to be defeated to obtain the Song of Storms check

VincentsSin replied:
Shifted fire arrows down to "Song of Storms" rule.

CelestialKitsune commented on worlds/mm-recomp/Rules.py:
```diff
@@ -296,14 +330,191 @@ def get_location_rules(player):
+        "Graveyard Day 1 Chest":
+            lambda state: state.has("Captain's Hat", player) and can_smack_hard(state, player) and can_use_fire_arrows(state, player),
+        "Graveyard Day 2 Chest":    
+            lambda state: state.has("Captain's Hat", player) and can_smack_hard(state, player),
```
Needs explosives as well

VincentsSin replied:
Added

CelestialKitsune commented on worlds/mm-recomp/Rules.py:
```diff
@@ -296,14 +330,191 @@ def get_location_rules(player):
+            lambda state: state.has("Captain's Hat", player) and can_smack_hard(state, player),
+            
+        "Ikana Canyon Pamela's Father":
+            lambda state: can_play_song("Song of Healing", state, player),
```
also needs Song of Storms, to get rid of the Gibdos around the Muic Box, AND either Bombs (or Blast Mask) or Stone Mask:
- **Stone Mask** allowing you to bypass Pamela.
- **Bombs** allowing you to force her out of the house, so you can sneak in.

VincentsSin replied:
Rule modified to `can_play_song("Song of Healing", state, player) and can_play_song("Song of Storms", state, player) and (has_explosives(state, player) or state.has("Stone Mask", player,)),`

VincentsSin committed:
Update worlds/mm-recomp/Rules.py

Co-authored-by: Hazel <30158363+CelestialKitsune@users.noreply.github.com>

VincentsSin committed:
Update worlds/mm-recomp/Rules.py

Co-authored-by: Hazel <30158363+CelestialKitsune@users.noreply.github.com>

VincentsSin committed:
Update worlds/mm-recomp/Rules.py

Co-authored-by: Hazel <30158363+CelestialKitsune@users.noreply.github.com>

VincentsSin committed:
Updated Rules

VincentsSin committed:
Updated Regions

Added "Pirates' Fortress Sewers" and updated "Stone Tower (Inverted)" to be a sub-region to "Stone Tower"

LittleCube-hax commented on worlds/mm-recomp/Rules.py:
```diff
@@ -2,11 +2,6 @@
-def universal_item_rule(item):
-    pass
-
-
```
lol i need this, i'm gonna use it for keysanity

VincentsSin replied:
That's not even in the copy I originally downloaded, so however that got deleted wasn't by me. But can add it back in for ya

LittleCube-hax commented on worlds/mm-recomp/Regions.py:
```diff
@@ -10,14 +10,39 @@ class MMRRegionData(NamedTuple):
+    "Ikana Canyon": MMRRegionData(["Secret Shrine", "Beneath the Well", "Ikana Castle", "Stone Tower"]),
+    "Secret Shrine": MMRRegionData([]),
+    "Beneath the Well": MMRRegionData([]),
+    "Ikana Castle": MMRRegionData(["Beneath the Well]),
```
there's a quote missing here

LittleCube-hax replied:
Come to think of it, it would probably make more sense to go the other direction, wouldn't it? Give Beneath the Well an entrance into Ikana Castle, since that's the direction the player normally goes anyway.

VincentsSin committed:
Updated Rules

Added and modified from suggestions

VincentsSin committed:
Update Regions

`Goron Shrine` added as sub-region to `Goron Village`.

Swapped the sub-region set for `Beneath the Well` and `Ikana Castle`. Also added `Inside Ikana Castle` as a sub-region to `Ikana Castle`.

VincentsSin committed:
Updated Rules

Added locations and logic for Stone Tower Temple (Normal and Inverted)
